File: internal/api/jdwp/referencetype.go
package jdwp

import (
	"fmt"

	"cli-debugger/pkg/errors"
)

// ReferenceType Command Set Implementation
// ReferenceType command constants are already defined in protocol.go

// FieldInfo Field Information
type FieldInfo struct {
	FieldID   string
	Name      string
	Signature string
	Modifiers int
}

// MethodInfo Method Information
type MethodInfo struct {
	MethodID  string
	Name      string
	Signature string
	Modifiers int
}

// Signature Get class signature
func (c *Client) Signature(refTypeID string) (string, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, referenceTypeCommandSignature, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class signature")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class signature")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class signature failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	signature, err := reader.readString()
	if err != nil {
		return "", errors.WrapProtocolError(err, errors.ErrInvalidResponse, "Failed to read class signature")
	}

	return signature, nil
}

// Fields Get class fields
func (c *Client) Fields(refTypeID string) ([]*FieldInfo, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, referenceTypeCommandFields, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class fields")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class fields")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class fields failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	declared := reader.readInt()

	fields := make([]*FieldInfo, 0, declared)
	for i := 0; i < declared; i++ {
		fieldID := reader.readID(c.idsizes.FieldIDSize)
		name, _ := reader.readString()
		signature, _ := reader.readString()
		modifiers := reader.readInt()

		fields = append(fields, &FieldInfo{
			FieldID:   fieldID,
			Name:      name,
			Signature: signature,
			Modifiers: modifiers,
		})
	}

	return fields, nil
}

// Methods Get class methods
func (c *Client) Methods(refTypeID string) ([]*MethodInfo, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, referenceTypeCommandMethods, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class methods")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class methods")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class methods failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	declared := reader.readInt()

	methods := make([]*MethodInfo, 0, declared)
	for i := 0; i < declared; i++ {
		methodID := reader.readID(c.idsizes.MethodIDSize)
		name, _ := reader.readString()
		signature, _ := reader.readString()
		modifiers := reader.readInt()

		methods = append(methods, &MethodInfo{
			MethodID:  methodID,
			Name:      name,
			Signature: signature,
			Modifiers: modifiers,
		})
	}

	return methods, nil
}

// SourceFile Get source file name
func (c *Client) SourceFile(refTypeID string) (string, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, referenceTypeCommandSourceFile, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get source file")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get source file")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get source file failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	sourceFile, err := reader.readString()
	if err != nil {
		return "", errors.WrapProtocolError(err, errors.ErrInvalidResponse, "Failed to read source file")
	}

	return sourceFile, nil
}

// GetValues Get static field values
func (c *Client) GetValues(refTypeID string, fieldIDs []string) ([]interface{}, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)

	data = append(data, 0, 0, 0, byte(len(fieldIDs)))
	for _, fieldID := range fieldIDs {
		data = append(data, encodeID(fieldID, c.idsizes.FieldIDSize)...)
	}

	packet := createCommandPacketWithData(referenceTypeCommandSet, 6, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get static field values")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get static field values")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get static field values failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	values := make([]interface{}, 0, len(fieldIDs))
	for i := 0; i < len(fieldIDs); i++ {
		tag := reader.readByte()
		value, _ := reader.readValue(tag)
		values = append(values, value)
	}

	return values, nil
}

// GetValuesWithTags Get static field values with tags
func (c *Client) GetValuesWithTags(refTypeID string, fieldIDs []string) ([]byte, []interface{}, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)

	data = append(data, 0, 0, 0, byte(len(fieldIDs)))
	for _, fieldID := range fieldIDs {
		data = append(data, encodeID(fieldID, c.idsizes.FieldIDSize)...)
	}

	packet := createCommandPacketWithData(referenceTypeCommandSet, 6, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get static field values")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get static field values")
	}

	if reply.ErrorCode != 0 {
		return nil, nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get static field values failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	tags := make([]byte, 0, len(fieldIDs))
	values := make([]interface{}, 0, len(fieldIDs))
	for i := 0; i < len(fieldIDs); i++ {
		tag := reader.readByte()
		tags = append(tags, tag)
		value, _ := reader.readValue(tag)
		values = append(values, value)
	}

	return tags, values, nil
}

// ClassFileVersion Get class file version
func (c *Client) ClassFileVersion(refTypeID string) (int, int, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, 17, data)
	if err := c.sendPacket(packet); err != nil {
		return 0, 0, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class file version")
	}

	reply, err := c.readReply()
	if err != nil {
		return 0, 0, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class file version")
	}

	if reply.ErrorCode != 0 {
		return 0, 0, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class file version failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	majorVersion := reader.readInt()
	minorVersion := reader.readInt()

	return majorVersion, minorVersion, nil
}

// ConstantPool Get class constant pool
func (c *Client) ConstantPool(refTypeID string) ([]byte, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, 18, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class constant pool")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class constant pool")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class constant pool failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	constantPoolCount := reader.readInt()

	constantPool := make([]byte, 0, constantPoolCount)
	for i := 0; i < constantPoolCount; i++ {
		b := reader.readByte()
		constantPool = append(constantPool, b)
	}

	return constantPool, nil
}

// Instances Get class instances
func (c *Client) Instances(refTypeID string, maxInstances int) ([]string, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)

	maxInstancesBytes := make([]byte, 4)
	maxInstancesBytes[0] = byte(maxInstances >> 24)
	maxInstancesBytes[1] = byte(maxInstances >> 16)
	maxInstancesBytes[2] = byte(maxInstances >> 8)
	maxInstancesBytes[3] = byte(maxInstances)

	data = append(data, maxInstancesBytes...)

	packet := createCommandPacketWithData(referenceTypeCommandSet, 16, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class instances")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class instances")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class instances failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	instances := make([]string, 0, maxInstances)
	instanceCount := reader.readInt()
	for i := 0; i < instanceCount; i++ {
		instance := reader.readID(c.idsizes.ObjectIDSize)
		instances = append(instances, instance)
	}

	return instances, nil
}

// ClassLoader Get class loader
func (c *Client) ClassLoader(refTypeID string) (string, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, 19, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class loader")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get class loader")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get class loader failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	classLoaderID := reader.readID(c.idsizes.ObjectIDSize)

	return classLoaderID, nil
}

// Module Get module
func (c *Client) Module(refTypeID string) (string, error) {
	data := encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)

	packet := createCommandPacketWithData(referenceTypeCommandSet, 20, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get module failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	moduleID := reader.readID(c.idsizes.ObjectIDSize)

	return moduleID, nil
}

File: src/protocol/jdwp/reference-type.ts
/**
 * ReferenceType Command Set Implementation
 */

import {
  CommandSet,
  ReferenceTypeCommand,
  type IDSizes,
  type FieldInfo,
  type MethodInfo,
} from "./protocol/index.js";
import {
  createCommandPacketWithData,
  encodeID,
  encodeUint32,
  encodeValue,
} from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get class signature
 */
export async function getSignature(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<string> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Signature,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get signature failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readString();
}

/**
 * Get class fields
 */
export async function getFields(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<FieldInfo[]> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Fields,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get fields failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const declared = reader.readInt();

  const fields: FieldInfo[] = [];
  for (let i = 0; i < declared; i++) {
    const fieldID = reader.readID(executor.idSizes.fieldIDSize);
    const name = reader.readString();
    const signature = reader.readString();
    const modifiers = reader.readInt();

    fields.push({ fieldID, name, signature, modifiers });
  }

  return fields;
}

/**
 * Get class methods
 */
export async function getMethods(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<MethodInfo[]> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Methods,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get methods failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const declared = reader.readInt();

  const methods: MethodInfo[] = [];
  for (let i = 0; i < declared; i++) {
    const methodID = reader.readID(executor.idSizes.methodIDSize);
    const name = reader.readString();
    const signature = reader.readString();
    const modifiers = reader.readInt();

    methods.push({ methodID, name, signature, modifiers });
  }

  return methods;
}

/**
 * Get source file name
 */
export async function getSourceFile(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<string> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.SourceFile,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get source file failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readString();
}

/**
 * Get static field values
 */
export async function getStaticFieldValues(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  fieldIDs: string[],
): Promise<unknown[]> {
  const parts: Buffer[] = [
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeUint32(fieldIDs.length),
  ];

  for (const fieldID of fieldIDs) {
    parts.push(encodeID(fieldID, executor.idSizes.fieldIDSize));
  }

  const data = Buffer.concat(parts);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.GetValues,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get static field values failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const values: unknown[] = [];

  for (let i = 0; i < fieldIDs.length; i++) {
    const { value } = reader.readTaggedValue(executor.idSizes.objectIDSize);
    values.push(value);
  }

  return values;
}

/**
 * Get static field values with tags
 */
export async function getValuesWithTags(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  fieldIDs: string[],
): Promise<{ tags: number[]; values: unknown[] }> {
  const parts: Buffer[] = [
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeUint32(fieldIDs.length),
  ];

  for (const fieldID of fieldIDs) {
    parts.push(encodeID(fieldID, executor.idSizes.fieldIDSize));
  }

  const data = Buffer.concat(parts);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.GetValues,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get static field values failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const tags: number[] = [];
  const values: unknown[] = [];

  for (let i = 0; i < fieldIDs.length; i++) {
    const { tag, value } = reader.readTaggedValue(executor.idSizes.objectIDSize);
    tags.push(tag);
    values.push(value);
  }

  return { tags, values };
}

/**
 * Set static field value
 */
export async function setStaticFieldValue(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  fieldID: string,
  value: unknown,
): Promise<void> {
  const parts: Buffer[] = [
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeUint32(1),
    encodeID(fieldID, executor.idSizes.fieldIDSize),
    encodeValue(value, executor.idSizes.objectIDSize),
  ];

  const data = Buffer.concat(parts);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.SetValues,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Set static field value failed: ${reply.message}`,
    );
  }
}

/**
 * Get class status
 */
export async function getStatus(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<number> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Status,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get class status failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readInt();
}

/**
 * Get implemented interfaces
 */
export async function getInterfaces(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<string[]> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Interfaces,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get interfaces failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const count = reader.readInt();

  const interfaces: string[] = [];
  for (let i = 0; i < count; i++) {
    interfaces.push(reader.readID(executor.idSizes.referenceTypeIDSize));
  }

  return interfaces;
}

/**
 * Get class object
 */
export async function getClassObject(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<string> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.ClassObject,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get class object failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readID(executor.idSizes.objectIDSize);
}

/**
 * Get class instances
 */
export async function getInstances(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  maxInstances: number,
): Promise<string[]> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeUint32(maxInstances),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.Instances,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get instances failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const count = reader.readInt();

  const instances: string[] = [];
  for (let i = 0; i < count; i++) {
    instances.push(reader.readID(executor.idSizes.objectIDSize));
  }

  return instances;
}

/**
 * Get class file version
 */
export async function getClassFileVersion(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<{ majorVersion: number; minorVersion: number }> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.ClassFileVersion,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get class file version failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return {
    majorVersion: reader.readInt(),
    minorVersion: reader.readInt(),
  };
}

/**
 * Get class loader
 */
export async function getClassLoader(
  executor: JDWPCommandExecutor,
  refTypeID: string,
): Promise<string> {
  const data = encodeID(refTypeID, executor.idSizes.referenceTypeIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ReferenceType,
    ReferenceTypeCommand.ClassLoader,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get class loader failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readID(executor.idSizes.objectIDSize);
}
